text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,988 @@
+        signature: RingSignature,
+        ring_proof: RingProof,
+        outputs: SEQUENCE_OF VrfOutput
+    }
+```
+
+- `signature`: represents the actual signature (opaque).
+- `ring_proof`: denotes the proof of ring membership.
+- `outputs`: sequence of `VrfOutput` objects corresponding to the `VrfInput` values.
+
+Helper function to create a `RingVrfSignature` from `VrfSignatureData`:
+
+```rust
+    fn ring_vrf_sign(
+        secret: BandersnatchSecretKey,
+        signature_data: VrfSignatureData,
+        prover: RingProver
+    ) -> RingVrfSignature;
+```
+
+Helper function for validating the signature and returning a Boolean value
+indicating the validity of the signature (`True` if it's valid).
+It's important to note that this function does not require the signer's public key.
+
+```rust
+    fn ring_vrf_verify(
+        signature: RingVrfSignature,
+        verifier: RingVerifier
+    ) -> Boolean;
+```
+
+In this document, the types `BandersnatchSecretKey`, `BandersnatchPublicKey`,
+`RingSignature`, `RingProof`, `RingProver` and `RingVerifier` are intentionally
+left undefined as are not relevant. Their definitions can be found in the
+`bandersnatch_vrfs` reference implementation.
+
+
+## 6. Sassafras Protocol
+
+### 6.1. Epoch's First Block
+
+The first block produced for an epoch `N` is required to include the descriptor
+for the next epoch `N+1`.
+
+The descriptor of next epoch is the `NextEpochDescriptor`.
+   
+```rust
+    AuthorityId ::= BandersnatchPublicKey;
+
+    Randomness ::= OCTET_STRING(SIZE(32));
+
+    NextEpochDescriptor ::= SEQUENCE {
+        authorities: SEQUENCE_OF AuthorityId,
+        randomness: Randomness,
+        configuration: ProtocolConfiguration OPTIONAL
+    }
+```
+
+- `authorities`: list of authorities for the next epoch.
+- `randomness`: randomness value associated to the next epoch.
+- `configuration`: next epoch optional protocol configuration.
+
+The `NextEpochDescriptor` must be `SCALE` encoded and embedded in the block
+header digest data.
+
+The identifier for the digest element is `BYTES("SASS")`.
+
+**Security Consideration**: Instances of `NextEpochDescriptor` are generated through
+on-chain code whenever a block is identified as the first of an epoch.
+Consequently, every node executing the block has the capability to verify if
+the descriptor generated during block execution matches the one produced by the
+block author, which is stored in the digest data.
+
+#### 6.1.1. Epoch Randomness
+
+Each epoch has an associated randomness value defined by the
+`NextEpochDescriptor` `Randomness` element.
+
+The first block of epoch `N` contains the randomness associated to the next epoch `N+1`.
+
+Randomness for epoch `N+1` is computed using the VRF output values which are associated
+to each block of epoch `N-1`.
+
+Assuming block `Bi` is produced during epoch `N-1`. Then the randomness for epoch `N+1`
+is incrementally generated as `RandomnessAccumulator = Blake2(Bi.VrfOut)'`.
```

Comment:
`Bi.RandomnessAccumulator = Blake2(Bi.parent.RandomnessAccumulator, Bi.claim.vrf_output())` or something like that.